wazp/callbacks.py

```diff
@@ -415,6 +420,612 @@ def modify_rows_selection(
+        """
+        # Get what triggered the callback
+        trigger = dash.callback_context.triggered[0]["prop_id"]
+        # Get the paths to the video and metadata files
+        video_path_pl = pl.Path(video_path)
+        video_name = video_path_pl.name
+        metadata_filepath = video_path_pl.with_suffix(".metadata.yaml")
+
+        # Check if the metadata file exists
+        # and load any previously saved ROIs
+        if metadata_filepath.exists():
+            with open(metadata_filepath, "r") as yaml_file:
+                metadata = yaml.safe_load(yaml_file)
+                if "ROIs" in metadata.keys():
+                    saved_rois = metadata["ROIs"]
+                else:
+                    saved_rois = []
+        else:
+            alert_msg = f"Could not find {metadata_filepath.name}"
+            alert_color = "danger"
+            return dash.no_update, alert_msg, alert_color, True
+
+        # Get the stored ROI shapes for this video
+        if video_name in roi_storage.keys():
+            roi_shapes = roi_storage[video_name]["shapes"]
+            rois_to_save = [
+                utils.stored_shape_to_yaml_entry(shape) for shape in roi_shapes
+            ]
+        else:
+            rois_to_save = []
+
+        if rois_to_save:
+            if trigger == "roi-storage.data" and rois_to_save == saved_rois:
+                # This means that the ROIs have just been loaded
+                alert_msg = f"Loaded ROIs from {metadata_filepath.name}"
+                alert_color = "success"
+                return dash.no_update, alert_msg, alert_color, True
+            elif trigger == "roi-storage.data" and rois_to_save != saved_rois:
+                # This means that the ROIs have been modified
+                # in respect to the metadata file
+                alert_msg = "Detected unsaved changes to ROIs."
+                alert_color = "warning"
+                return dash.no_update, alert_msg, alert_color, True
+            else:
+                if save_clicks > 0:
+                    # This means that the user wants to save the ROIs
+                    # This overwrites any previously saved ROIs
+                    with open(metadata_filepath, "w") as yaml_file:
+                        metadata["ROIs"] = rois_to_save
+                        yaml.safe_dump(metadata, yaml_file)
+                    alert_msg = f"Saved ROIs to {metadata_filepath.name}"
+                    alert_color = "success"
+                    return (
+                        metadata_filepath.as_posix(),
+                        alert_msg,
+                        alert_color,
+                        True,
+                    )
+                else:
+                    return dash.no_update, dash.no_update, dash.no_update, True
+
+        else:
+            alert_msg = "No ROIs to save."
+            alert_color = "light"
+            return dash.no_update, alert_msg, alert_color, True
```

Comment:
I don't think this is your fault, but these `get_flibble_callbacks` functions get a bit [long](https://refactoring.guru/smells/long-method). Is there a nice way to refactor? (Maybe not for now)

Reply:
In terms of callbacks inputs/output, I already tried to simplify things as far as I could.
But I think there is still room for refactoring small things (such as above) into separate utility functions.
I will leave that for the future though